File: all-submissions/0323-number-of-connected-components-in-an-undirected-graph/solution.java
class Solution {

    Map<Integer, List<Integer>> graph = new HashMap<>();

    public int countComponents(int n, int[][] edges) {

        //build the graph
        for(int[] edge: edges){
            int x = edge[0];
            int y = edge[1];
            if(!graph.containsKey(x)) {
                graph.put(x, new ArrayList<>());
            }
            if(!graph.containsKey(y)) {
                graph.put(y, new ArrayList<>());
            }
            graph.get(x).add(y);
            graph.get(y).add(x);
        }

        //DFS
        int count = 0;
        boolean[] seen = new boolean[n];
        for(int i=0;i<n;i++){
            if(!graph.containsKey(i)) {
                count++;
                seen[i] = true;
            } else
            if(!seen[i]){
                count++;
                seen[i] = true;
                dfs(i, seen);
            }
        }
        return count;
    }

    public void dfs(int index, boolean[] seen) {
        List<Integer> neighbors = graph.get(index);
        for(int neighbor: neighbors) {
            if(!seen[neighbor]) {
                seen[neighbor] = true;
                dfs(neighbor, seen);
            }
        }
    }
}

File: all-submissions/0323-number-of-connected-components-in-an-undirected-graph/solution.py
class Solution:
    def countComponents(self, n: int, edges: List[List[int]]) -> int:
        adj = defaultdict(list)
        for edge in edges:
            adj[edge[0]].append(edge[1])
            adj[edge[1]].append(edge[0])

        def dfs(i):
            for edge in adj[i]:
                if edge not in visited:
                    visited.add(edge)
                    dfs(edge)
        count = 0
        visited = set()
        for i in range(n):
            if i not in visited:
                count +=1
                visited.add(i)
                dfs(i)

        return count
